Services/TMS/Booking/classes/class.ilTMSBookingActions.php

```diff
@@ -23,29 +23,131 @@ public function bookUser($crs_ref_id, $user_id) {
 		$user = ilObjectFactory::getInstanceByObjId($user_id);
 		assert('$user instanceof \ilObjUser');
 
-		$this->maybeMakeCourseMember($course, $user);
+		return $this->maybeMakeCourseMember($course, $user);
+	}
+
+	/**
+	 * Removes a user from course or waiting list
+	 *
+	 * @param	int		$crs_ref_id
+	 * @param	int		$user_id
+	 *
+	 * @return string | null
+	 */
+	public function stornoUser($crs_ref_id, $user_id) {
+		require_once("Modules/Course/classes/class.ilCourseParticipants.php");
+		require_once("Services/Membership/classes/class.ilWaitingList.php");
+		$course = ilObjectFactory::getInstanceByRefId($crs_ref_id);
+		if(ilCourseParticipants::_isParticipant($course->getRefId(), $user_id)) {
+			$course->getMemberObject()->delete($user_id);
+			return Booking\Actions::STATE_REMOVED;
+		}
+
+		$crs_id = $course->getId();
+		if(ilWaitingList::_isOnList($user_id, $crs_id)) {
+			ilWaitingList::deleteUserEntry($user_id, $crs_id);
+			return Booking\Actions::STATE_REMOVED;
+		}
 
-		return Booking\Actions::STATE_BOOKED;
+		return null;
```

Comment:
Wie kann das Ding in diesem Zustand sein? Imo sollte hier eine Exception geworfen werden, oder?

Reply:
Die Fehlermeldung übernimmt der Step. Ich finde der soll entscheiden was bei return null passiert. Immerhin beauftragt er ja auch alles.